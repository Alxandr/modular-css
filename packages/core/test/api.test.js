"use strict";

const path = require("path");

const namer = require("test-utils/namer.js");
const { find, cwd, relative } = require("test-utils/fixtures.js");

const Processor = require("../processor.js");

describe("/processor.js", () => {
    describe("API", () => {
        let processor;

        beforeEach(() => {
            processor = new Processor({
                namer,
                cwd,
            });
        });

        it("should be a function", () => {
            expect(typeof Processor).toBe("function");
        });
        
        describe(".string()", () => {
            it("should process a string", async () => {
                const result = await processor.string(
                    "./simple.css", ".wooga { }"
                );

                expect(result.exports).toMatchSnapshot();
                expect(result.details.exports).toMatchSnapshot();
                expect(result.details.text).toMatchSnapshot();
                expect(result.details.processed.root.toResult().css).toMatchSnapshot();
            });
        });
        
        describe(".file()", () => {
            it("should process a relative file", async () => {
                const result = await processor.file(find("simple.css"));

            
                expect(result.exports).toMatchSnapshot();
                expect(result.details.exports).toMatchSnapshot();
                expect(result.details.text).toMatchSnapshot();
                expect(result.details.processed.root.toResult().css).toMatchSnapshot();
            });

            it("should process an absolute file", async () => {
                const result = await processor.file(find("simple.css"));

                expect(result.exports).toMatchSnapshot();
                expect(result.details.exports).toMatchSnapshot();
                expect(result.details.text).toMatchSnapshot();
                expect(result.details.processed.root.toResult().css).toMatchSnapshot();
            });
        });
        
        describe(".remove()", () => {
            it("should remove a relative file", async () => {
                await processor.string(
                    "./simple.css",
                    ".wooga { }"
                );

                processor.remove("./simple.css");
                    
                expect(processor.dependencies().map(relative)).toMatchSnapshot();
            });

            it("should remove an absolute file", async () => {
                await processor.string(
                    find("simple.css"),
                    ".wooga { }"
                );

                processor.remove(find("simple.css"));
                    
                expect(processor.dependencies().map(relative)).toMatchSnapshot();
            });
            
            it("should remove multiple files", async () => {
                await processor.string("./a.css", ".a { }");
                await processor.string("./b.css", ".b { }");
                await processor.string("./c.css", ".c { }");
                
                processor.remove([
                    "./a.css",
                    "./b.css",
                ]);
                    
                expect(processor.dependencies().map(relative)).toMatchSnapshot();
            });
            
            it("should return an array of removed files", async () => {
                await processor.string("./a.css", ".a { }");
                await processor.string("./b.css", ".b { }");
                await processor.string("./c.css", ".c { }");
               
                expect(
                    processor.remove([
                        "./a.css",
                        "./b.css",
                    ]).map(relative)
                ).toMatchSnapshot();
            });
        });
        
        describe(".dependencies()", () => {
            it("should return the dependencies of the specified file", async () => {
                await processor.file(find("start.css"));

                const dependencies = processor.dependencies(find("start.css")).map(relative);
                
                expect(dependencies).toMatchSnapshot();
            });
            
            it("should return the overall order of dependencies if no file is specified", async () => {
                await processor.file(find("start.css"));

                const dependencies = processor.dependencies().map(relative);

                expect(dependencies).toMatchSnapshot();
            });
        });

        describe(".dependents()", () => {
            it("should return the dependents of the specified file", async () => {
                await processor.file(find("start.css"));

                const dependents = processor.dependents(find("local.css")).map(relative);

                expect(dependents).toMatchSnapshot();
            });
            
            it("should throw if no file is passed", async () => {
                await processor.file(find("start.css"));

                expect(() => processor.dependents()).toThrowErrorMatchingSnapshot();
            });
        });
        
        describe(".output()", () => {
            it("should return a postcss result", async () => {
                await processor.file(find("start.css"));
                
                const result = await processor.output();
                
                expect(result.css).toMatchSnapshot();
            });
            
            it("should generate css representing the output from all added files", async () => {
                await processor.file(find("start.css"));
                await processor.file(find("simple.css"));
                
                const result = await processor.output();

                expect(result.css).toMatchSnapshot();
            });

            it("should avoid duplicating files in the output", async () => {
                await processor.file(find("start.css"));
                await processor.file(find("local.css"));
                
                const result = await processor.output();

                expect(result.css).toMatchSnapshot();
            });
            
            it("should generate a JSON structure of all the compositions", async () => {
                await processor.file(find("start.css"));
                
                const result = await processor.output();

                expect(result.compositions).toMatchSnapshot();
            });
            
            it("should order output by dependencies, then alphabetically", async () => {
                await processor.file(find("start.css"));
                await processor.file(find("local.css"));
                await processor.file(find("composes.css"));
                await processor.file(find("deep.css"));
                
                const result = await processor.output();

                expect(result.css).toMatchSnapshot();
            });

            it("should support returning output for specified relative files", async () => {
                await processor.file(find("start.css"));
                await processor.file(find("local.css"));
                
                const result = await processor.output({
                    files : [
                        find("start.css"),
                    ],
                });

                expect(result.css).toMatchSnapshot();
            });

            it("should support returning output for specified absolute files", async () => {
                await processor.file(find("start.css"));
                await processor.file(find("local.css"));
                
                const result = await processor.output({
                    files : [
                        find("start.css"),
                    ],
                });

                expect(result.css).toMatchSnapshot();
            });

            it("should reject if called before input has been processed", () => {
                processor.file(find("start.css"));

                return expect(processor.output()).rejects.toMatchSnapshot();
            });

            it("should allow for seperate source map output", async () => {
                await processor.file(find("start.css"));
                
                const result = await processor.output({
                    to  : path.join(cwd, "./to.css"),
                    map : {
                        inline : false,
                    },
                });

                expect(result.map).toMatchSnapshot();
            });
        });

        describe("._resolve()", () => {
            it("should run resolvers until a match is found", () => {
                var ran = false;

                processor = new Processor({
                    cwd,
                    resolvers : [
                        () => {
                            ran = true;
                        },
                    ],
                });
                
                expect(relative(
                    processor._resolve(
                        find("start.css"),
                        "./local.css"
                    ),
                ))
                .toMatchSnapshot();

                expect(ran).toBeTruthy();
            });

            it("should fall back to a default resolver", () => {
                processor = new Processor({
                    cwd,
                    resolvers : [
                        () => undefined,
                    ],
                });
                
                expect(relative(
                    processor._resolve(
                        find("start.css"),
                        "./local.css"
                    ),
                ))
                .toMatchSnapshot();
            });
        });
    });
});
