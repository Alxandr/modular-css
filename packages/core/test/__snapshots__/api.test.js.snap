// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/processor.js API ._resolve() should fall back to a default resolver 1`] = `
Array [
  "fixtures/local.css",
]
`;

exports[`/processor.js API ._resolve() should run resolvers until a match is found 1`] = `
Array [
  "fixtures/local.css",
]
`;

exports[`/processor.js API .dependencies() should return the dependencies of the specified file 1`] = `
Array [
  "fixtures/folder/folder.css",
  "fixtures/local.css",
]
`;

exports[`/processor.js API .dependencies() should return the overall order of dependencies if no file is specified 1`] = `
Array [
  "fixtures/folder/folder.css",
  "fixtures/local.css",
  "fixtures/start.css",
]
`;

exports[`/processor.js API .dependents() should return the dependents of the specified file 1`] = `
Array [
  "fixtures/start.css",
]
`;

exports[`/processor.js API .dependents() should throw if no file is passed 1`] = `"Must provide a file to processor.dependants()"`;

exports[`/processor.js API .file() should process a relative file 1`] = `
Object {
  "fooga": Array [
    "fooga",
  ],
  "str": "\\"string\\"",
}
`;

exports[`/processor.js API .file() should process a relative file 2`] = `
Object {
  "fooga": Array [
    "fooga",
  ],
  "str": "\\"string\\"",
}
`;

exports[`/processor.js API .file() should process a relative file 3`] = `
"@value str: \\"string\\";

.fooga {
    color: red;
}
"
`;

exports[`/processor.js API .file() should process a relative file 4`] = `
".fooga {
    color: red;
}
"
`;

exports[`/processor.js API .file() should process an absolute file 1`] = `
Object {
  "fooga": Array [
    "fooga",
  ],
  "str": "\\"string\\"",
}
`;

exports[`/processor.js API .file() should process an absolute file 2`] = `
Object {
  "fooga": Array [
    "fooga",
  ],
  "str": "\\"string\\"",
}
`;

exports[`/processor.js API .file() should process an absolute file 3`] = `
"@value str: \\"string\\";

.fooga {
    color: red;
}
"
`;

exports[`/processor.js API .file() should process an absolute file 4`] = `
".fooga {
    color: red;
}
"
`;

exports[`/processor.js API .output() should allow for seperate source map output 1`] = `
Object {
  "file": "to.css",
  "mappings": "AAAA,gCAAC;AACD,UAAU,YAAY,EAAE;ACDxB,wBAAC;AAID,SAAS,kBAAkB,EAAE;ACJ7B,wBAAC;AAED,SDFA,WAAgB,CAChB,iBAAiB,ECCsB;AACvC,SFHA,wBAAqB,EEGe",
  "names": Array [],
  "sources": Array [
    "fixtures/folder/folder.css",
    "fixtures/local.css",
    "fixtures/start.css",
  ],
  "sourcesContent": Array [
    "@value folder: white;
.folder { margin: 2px; }
",
    "@value one: red;
@value two: blue;
@value folder from \\"./folder/folder.css\\";

.booga { background: green; }
.looga { composes: booga; }
",
    "@value one, two, folder from \\"./local.css\\";
.wooga { composes: booga from \\"./local.css\\"; }
.booga { color: one; background: two; }
.tooga { border: 1px solid folder; }
",
  ],
  "version": 3,
}
`;

exports[`/processor.js API .output() should avoid duplicating files in the output 1`] = `
"/* fixtures/folder/folder.css */
.folder { margin: 2px; }
/* fixtures/local.css */
.booga { background: green; }
/* fixtures/start.css */
.booga { color: red; background: blue; }
.tooga { border: 1px solid white; }
"
`;

exports[`/processor.js API .output() should generate a JSON structure of all the compositions 1`] = `
Object {
  "fixtures/folder/folder.css": Object {
    "folder": "folder",
  },
  "fixtures/local.css": Object {
    "booga": "booga",
    "folder": "white",
    "looga": "booga looga",
    "one": "red",
    "two": "blue",
  },
  "fixtures/start.css": Object {
    "booga": "booga",
    "folder": "white",
    "one": "red",
    "tooga": "tooga",
    "two": "blue",
    "wooga": "booga wooga",
  },
}
`;

exports[`/processor.js API .output() should generate css representing the output from all added files 1`] = `
"/* fixtures/folder/folder.css */
.folder { margin: 2px; }
/* fixtures/simple.css */
.fooga {
    color: red;
}
/* fixtures/local.css */
.booga { background: green; }
/* fixtures/start.css */
.booga { color: red; background: blue; }
.tooga { border: 1px solid white; }
"
`;

exports[`/processor.js API .output() should order output by dependencies, then alphabetically 1`] = `
"/* fixtures/folder/folder.css */
.folder { margin: 2px; }
/* fixtures/folder/folder2.css */
.folder2 { color: green; }
/* fixtures/folder/subfolder/subfolder.css */
.subfolder { color: yellow; }
/* fixtures/composes.css */
/* fixtures/deep.css */
.deep {
    
    color: black;
}
/* fixtures/local.css */
.booga { background: green; }
/* fixtures/start.css */
.booga { color: red; background: blue; }
.tooga { border: 1px solid white; }
"
`;

exports[`/processor.js API .output() should reject if called before input has been processed 1`] = `[Error: File processing not complete]`;

exports[`/processor.js API .output() should return a postcss result 1`] = `
"/* fixtures/folder/folder.css */
.folder { margin: 2px; }
/* fixtures/local.css */
.booga { background: green; }
/* fixtures/start.css */
.booga { color: red; background: blue; }
.tooga { border: 1px solid white; }
"
`;

exports[`/processor.js API .output() should support returning output for specified absolute files 1`] = `
"/* fixtures/start.css */
.booga { color: red; background: blue; }
.tooga { border: 1px solid white; }
"
`;

exports[`/processor.js API .output() should support returning output for specified relative files 1`] = `
"/* fixtures/start.css */
.booga { color: red; background: blue; }
.tooga { border: 1px solid white; }
"
`;

exports[`/processor.js API .remove() should remove a relative file 1`] = `Array []`;

exports[`/processor.js API .remove() should remove an absolute file 1`] = `Array []`;

exports[`/processor.js API .remove() should remove multiple files 1`] = `
Array [
  "c.css",
]
`;

exports[`/processor.js API .remove() should return an array of removed files 1`] = `
Array [
  "a.css",
  "b.css",
]
`;

exports[`/processor.js API .string() should process a string 1`] = `
Object {
  "wooga": Array [
    "wooga",
  ],
}
`;

exports[`/processor.js API .string() should process a string 2`] = `
Object {
  "wooga": Array [
    "wooga",
  ],
}
`;

exports[`/processor.js API .string() should process a string 3`] = `".wooga { }"`;

exports[`/processor.js API .string() should process a string 4`] = `".wooga { }"`;
